### ISSUE TITLE
scipy.signal.detrend implementation in JAX

### ISSUE THREAD

**Issue by docyyz** [google/jax#3498]
HI, 

Would it be possible to implement scipy.signal.detrend in JAX?  I need to compute the gradient of a function which involves de-trending a signal computed from input parameters.

Thanks,
York

**Comment by jakevdp** [comment]
Here's a quick implementation (not comprehensively tested):
```python
import jax.numpy as jnp
import numpy as np

def detrend(data, axis=-1, type='linear', bp=0):
  if type not in ['constant', 'linear', 'c', 'l']:
    raise ValueError("Trend type must be 'linear' or 'constant'.")
  data = jnp.asarray(data)
  if type in ['constant', 'c']:
    return data - jnp.mean(data, axis, keepdims=True)
  else:
    N = data.shape[axis]
    # bp is static, so we use np operations to avoid pushing to device.
    bp = np.sort(np.unique(np.r_[0, bp, N]))
    if np.any(bp > N):
      raise ValueError("Breakpoints must be less than length of data along given axis.")
    data = jnp.moveaxis(data, axis, 0)
    shape = data.shape
    data = data.reshape(N, -1)
    for m in range(len(bp) - 1):
      Npts = bp[m + 1] - bp[m]
      A = jnp.vstack([jnp.ones(Npts), jnp.arange(1, Npts + 1) / Npts]).T
      sl = slice(bp[m], bp[m + 1])
      coef, resids, rank, s = jnp.linalg.lstsq(A, data[sl])
      data = data.at[sl].add(-jnp.dot(A, coef))
    return jnp.moveaxis(data.reshape(shape), 0, axis)
```
I can work on getting it into ``jax.scipy.signal`` this week.

**Comment by docyyz** [comment]
Thanks for the quick reply.  I tested the function with my data and it's very close to the result from using scipy.signal.detrend. It's a bit slower than the original scipy version, but I suppose wrapping it in jit will speed it up for repeated evaluations.
